src/main/java/app/linguistai/bmvp/service/gamification/StoreService.java

```diff
@@ -0,0 +1,206 @@
+                    .map(RUserItem::new)
+                    .collect(Collectors.toList());
+
+            // Build a response and return user's all items
+            RUserItems response = RUserItems.builder().userItems(userItems)
+                    .totalPages(userItemsPage.getTotalPages())
+                    .currentPage(userItemsPage.getNumber())
+                    .pageSize(userItemsPage.getSize())
+                    .build();
+            log.info("All user items are fetched for user with email {}", email);
+            return response;
+        } catch (NotFoundException e) {
+            log.error("Fetching user items failed since user does not exist for email {}", email);
+            throw e;
+        } catch (Exception e) {
+            log.error("Fetching user items failed for user with email {}", email, e);
+            throw new SomethingWentWrongException();
+        }
+    }
+
+    @Transactional
+    public RUserItem purchaseUserItem(String email, UUID storeItemId) throws Exception {
+        try {
+            User user = accountRepository.findUserByEmail(email)
+                    .orElseThrow(() -> new NotFoundException(User.class.getSimpleName(), true));
+
+            StoreItem storeItem = storeItemRepository.findById(storeItemId)
+                    .orElseThrow(() -> new NotFoundException(StoreItem.class.getSimpleName(), true));
+
+            // Check if the store item is enabled
+            if (!storeItem.isEnabled()) {
+                throw new ItemNotEnabledException();
+            }
+
+            transactionService.processTransaction(email, TransactionType.SHOP_SPEND, storeItem.getGemPrice());
+
+            // If the user already has the item, increment the quantity, otherwise create a new UserItem
+            UserItem userItem = userItemRepository.findByUserAndStoreItem(user, storeItem)
+                    .map(item -> {
+                        item.setQuantity(item.getQuantity() + 1);
+                        return item;
+                    })
+                    .orElseGet(() -> new UserItem(user, storeItem, 1));
+
+            // Save the user item
+            userItemRepository.save(userItem);
+
+            log.info("User {} purchased item {} successfully", email, storeItem.getType());
+            return RUserItem.builder().userItem(userItem).build();
+        } catch (NotFoundException e) {
+            if (e.getObject().equals(User.class.getSimpleName())) {
+                log.error("When purchasing an item, user is not found for email {}", email);
+            } else {
+                log.error("When purchasing an item, store item is not found for id {}", storeItemId);
+            }
+            throw e;
+        } catch (ItemNotEnabledException e) {
+            log.error("Store item with id {} is not enabled for purchase", storeItemId);
+            throw e;
+        }  catch (InsufficientGemsException e) {
```

Comment:
This log should already be thrown in `transactionService.processTransaction()`. Re-logging here may log the error twice. Can you please check? If so, I think it may be better to remove this catch block here, since the transaction functionality and the related logging responsibility is on `transactionService`.

Reply:
Yes, this is a duplicate log. I will remove the log but keep the catch block, since otherwise InsufficientGemsException is caught by `catch (Exception e)`, which returns a much more generic error message indicating there was a system error.